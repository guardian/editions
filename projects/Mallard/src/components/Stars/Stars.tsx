import React, { useMemo } from 'react';
import { StyleSheet, Text, View } from 'react-native';
import { color } from 'src/theme/color';
import { metrics } from 'src/theme/spacing';
import { getFont } from 'src/theme/typography';

const styles = StyleSheet.create({
	background: {
		backgroundColor: color.palette.highlight.main,
		padding: 0,
		paddingBottom: 2,
		paddingHorizontal: metrics.horizontal / 3,
	},
	text: {
		letterSpacing: 0.25,
		...getFont('icon', 1),
	},
	bottomLeft: {
		position: 'absolute',
		left: 0,
		bottom: 0,
	},
	inline: {
		flex: 0,
	},
});

const getRatingAsText = (rating: number) =>
	['☆', '☆', '☆', '☆', '☆'].map((s, index) => {
		if (index + 1 <= rating) {
			return '★';
		}
		if (index + 0.01 <= rating) {
			return '';
		}
		return s;
	});

const Stars = ({
	position,
	rating,
}: {
	position?: 'bottomLeft' | 'inline';
	rating: number;
}) => {
	const ratingAsText = useMemo(() => getRatingAsText(rating), [rating]);
	return (
		<View
			accessibilityLabel={`${rating.toString()} stars`}
			style={[styles.background, position && styles[position]]}
		>
			<Text style={styles.text} accessible={false}>
				{ratingAsText.join('')}
			</Text>
		</View>
	);
};

export { Stars };
