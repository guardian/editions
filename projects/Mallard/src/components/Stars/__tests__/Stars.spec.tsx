import React from 'react';
import TestRenderer from 'react-test-renderer';
import { Stars } from '../../../components/Stars/Stars';

const rating = 5;

describe('Stars', () => {
	it('should show a Stars with default styling', () => {
		const component = TestRenderer.create(
			<Stars rating={rating} />,
		).toJSON();
		expect(component).toMatchSnapshot();
	});

	it('should show a Stars with trailImage styling', () => {
		const component = TestRenderer.create(
			<Stars rating={rating} position="bottomLeft" />,
		).toJSON();
		expect(component).toMatchSnapshot();
	});

	it('should show a Stars with smallItems styling', () => {
		const component = TestRenderer.create(
			<Stars rating={rating} position="inline" />,
		).toJSON();
		expect(component).toMatchSnapshot();
	});
});
