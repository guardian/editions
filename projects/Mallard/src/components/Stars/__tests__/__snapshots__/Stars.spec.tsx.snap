// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stars should show a Stars with default styling 1`] = `
<View
  accessibilityLabel="5 stars"
  style={
    Array [
      Object {
        "backgroundColor": "#ffe500",
        "padding": 0,
        "paddingBottom": 2,
        "paddingHorizontal": 4.666666666666667,
      },
      undefined,
    ]
  }
>
  <Text
    accessible={false}
    style={
      Object {
        "fontFamily": "GuardianIcons-Regular",
        "fontSize": 20,
        "letterSpacing": 0.25,
        "lineHeight": 20,
      }
    }
  >
    ★★★★★
  </Text>
</View>
`;

exports[`Stars should show a Stars with smallItems styling 1`] = `
<View
  accessibilityLabel="5 stars"
  style={
    Array [
      Object {
        "backgroundColor": "#ffe500",
        "padding": 0,
        "paddingBottom": 2,
        "paddingHorizontal": 4.666666666666667,
      },
      Object {
        "flex": 0,
      },
    ]
  }
>
  <Text
    accessible={false}
    style={
      Object {
        "fontFamily": "GuardianIcons-Regular",
        "fontSize": 20,
        "letterSpacing": 0.25,
        "lineHeight": 20,
      }
    }
  >
    ★★★★★
  </Text>
</View>
`;

exports[`Stars should show a Stars with trailImage styling 1`] = `
<View
  accessibilityLabel="5 stars"
  style={
    Array [
      Object {
        "backgroundColor": "#ffe500",
        "padding": 0,
        "paddingBottom": 2,
        "paddingHorizontal": 4.666666666666667,
      },
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
      },
    ]
  }
>
  <Text
    accessible={false}
    style={
      Object {
        "fontFamily": "GuardianIcons-Regular",
        "fontSize": 20,
        "letterSpacing": 0.25,
        "lineHeight": 20,
      }
    }
  >
    ★★★★★
  </Text>
</View>
`;
